### PR TITLE
Limit $GENERATE range to 65535 steps

### LINKED ISSUE #1019: limit $GENERATE range to 2^16?
a $GENERATE range that is too large will cause a panic, it make sense to limit this to something reasonable. I propose to limit the number of steps to 65535, i.e.

`(stop - start) / step > 65535`, and error out if we hit this.

### PR DESCRIPTION
**Pull request by miekg**
Fixes #1019

### CODE REVIEW

File: generate_test.go [hunk 1]
```diff
@@ -229,3 +227,32 @@ func TestCrasherString(t *testing.T) {
...
 	}
 }
+
+func TestOutofRangeRange(t *testing.T) {
+	// outside limit
+	data := `$GENERATE 0-65537 dhcp A 10.0.0.1`
+	zp := NewZoneParser(strings.NewReader(data), ".", "test")
+	for _, ok := zp.Next(); ok; _, ok = zp.Next() {
+	}
+	if !strings.Contains(zp.Err().Error(), "dns: bad range in $GENERATE directive:") {
```
Comment (tmthrgd) [inline generate_test.go]: This should have `zp.Err() == nil || !strings.Contains(...)`.

File: generate.go [hunk 2]
```diff
@@ -72,6 +72,10 @@ func (zp *ZoneParser) generate(l lex) (RR, bool) {
 		s += l.token
 	}
 
+	if (end-start)/step > 65536 {
+		return zp.setParseError("bad range in $GENERATE directive", l)
```
Comment (tmthrgd) [inline generate.go]: Can we merge this with the `end < 0 || start < 0 || end < start` if-statement above?

**Comment by miekg** [comment]
done. PTAL

File: generate.go [hunk 3]
```diff
@@ -49,14 +49,14 @@ func (zp *ZoneParser) generate(l lex) (RR, bool) {
 	if err != nil {
 		return zp.setParseError("bad stop in $GENERATE range", l)
 	}
-	if end < 0 || start < 0 || end < start {
+	if end < 0 || start < 0 || end < start || (end-start)/step > 65535 {
```
Comment (chantra) [inline generate.go]: See my comment in https://github.com/miekg/dns/issues/1019#issuecomment-537989190 to align this with BIND behaviour. TL;DR both start and stop must be positive integers between 0 and (2^31)-1

File: generate_test.go [hunk 4]
```diff
@@ -205,27 +205,31 @@ $GENERATE 32-158 dhcp-${-32,4,d} A 10.0.0.$
...
 
-func TestCrasherString(t *testing.T) {
-	tests := []struct{
-	in  string
-	err string
-}{
-		{"$GENERATE 0-300103\"$$GENERATE 2-2", "dns: garbage after $GENERATE range: \"\\\"\" at line: 1:19"},
-		{"$GENERATE 0-5414137360", "dns: garbage after $GENERATE range: \"\\n\" at line: 1:22"},
-		{"$GENERATE       11522-3668518066406258", "dns: garbage after $GENERATE range: \"\\n\" at line: 1:38"},
-		{"$GENERATE 0-200\"(;00000000000000\n$$GENERATE 0-0", "dns: garbage after $GENERATE range: \"\\\"\" at line: 1:16"},
```
Comment (chantra) [inline generate_test.go]: some of those test were generating different errors, like timeout (I believe the second test was a timeout). Let me find the original error they were triggering.

File: generate_test.go [hunk 5]
```diff
@@ -205,27 +205,31 @@ $GENERATE 32-158 dhcp-${-32,4,d} A 10.0.0.$
...
+		t.Errorf("Expected nil error, got %s", zp.Err())
+	}
 
-func TestCrasherString(t *testing.T) {
-	tests := []struct{
-	in  string
-	err string
-}{
-		{"$GENERATE 0-300103\"$$GENERATE 2-2", "dns: garbage after $GENERATE range: \"\\\"\" at line: 1:19"},
-		{"$GENERATE 0-5414137360", "dns: garbage after $GENERATE range: \"\\n\" at line: 1:22"},
```
Comment (chantra) [inline generate_test.go]: original error: `program hanged (timeout 10 seconds)`

File: generate_test.go [hunk 6]
```diff
@@ -205,27 +205,31 @@ $GENERATE 32-158 dhcp-${-32,4,d} A 10.0.0.$
...
+	if zp.Err() != nil {
+		t.Errorf("Expected nil error, got %s", zp.Err())
+	}
 
-func TestCrasherString(t *testing.T) {
-	tests := []struct{
-	in  string
-	err string
-}{
-		{"$GENERATE 0-300103\"$$GENERATE 2-2", "dns: garbage after $GENERATE range: \"\\\"\" at line: 1:19"},
```
Comment (chantra) [inline generate_test.go]: runtime: goroutine stack exceeds 1000000000-byte limit fatal error: stack overflow

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Limit $GENERATE ran..." ]
>chantra commented on this pull request.
>
>
>
>>
>-func TestCrasherString(t *testing.T) {
>-	tests := []struct{
>-	in  string
>-	err string
>-}{
>-		{"$GENERATE 0-300103\"$$GENERATE 2-2", "dns: garbage after $GENERATE range: \"\\\"\" at line: 1:19"},
>-		{"$GENERATE 0-5414137360", "dns: garbage after $GENERATE range: \"\\n\" at line: 1:22"},
>
>original error: `program hanged (timeout 10 seconds)`

yes, with only 2^16-1 we don't hit this - I think the bind limits are a bit 
crazy

File: generate_test.go [hunk 7]
```diff
@@ -205,27 +205,31 @@ $GENERATE 32-158 dhcp-${-32,4,d} A 10.0.0.$
...
-		{"$GENERATE 0-5414137360", "dns: garbage after $GENERATE range: \"\\n\" at line: 1:22"},
-		{"$GENERATE       11522-3668518066406258", "dns: garbage after $GENERATE range: \"\\n\" at line: 1:38"},
```
Comment (chantra) [inline generate_test.go]: program hanged (timeout 10 seconds)

File: generate_test.go (hunk 4 above)
Comment (chantra) [inline generate_test.go]: panic: runtime error: index out of range [2048] with length 2048

**Comment by miekg** [comment]
@chantra happy to merge this as it now?

**Comment by chantra** [comment]
sure, I would be tempted to leave test case that catch garbage after the range/step token though :)

**Comment by miekg** [comment]
ok, adding that back in

**Pull request merged**